### ISSUE TITLE
Locking requirements for CI

### ISSUE THREAD

**Issue by tillahoffmann** [stan-dev/cmdstanpy#653]
The CI pipeline currently installs requirements from `requirements.txt`, `requirements-test.txt`, and `requirements-optional.txt`. These files list semantic requirements, e.g., `numpy`, but don't "lock" the specific version, e.g., `numpy==x.y.z`. During testing, this can lead to unexpected failures caused by changes in underlying packages rather than changes related to the code in this repository. 

For example, [this run](https://github.com/stan-dev/cmdstanpy/actions/runs/3895979307/jobs/6651952898) succeeded with `pylint==2.15.10`, but [this run](https://github.com/stan-dev/cmdstanpy/actions/runs/4168677373/jobs/7215767521) failed with `pylint==2.16.2`. This was due to the change https://github.com/PyCQA/pylint/issues/7690 released with [pylint 2.16.0](https://github.com/PyCQA/pylint/releases/tag/v2.16.0).

Keeping the requirements semantic for the actual package, but locking the requirements for CI for reproducible builds has worked well for me in the past. I'm happy to send a PR if this is of interest but wanted to check with you before making any changes.

**Comment by tillahoffmann** [comment]
That seems fair. 

I've been locking dependencies and then periodically updated the lock file. If the build for the commit that updates the lock file fails, I know that it is a dependency that is the underlying cause (rather than some other non-deterministic culprit). E.g., https://github.com/testcontainers/testcontainers-python/pull/274. But that may be overkill, and locking `pylint` gets the job done.

**Comment by WardBrian** [comment]
Do you know if we could have that periodic update be automated (e.g. by something like dependabot)?

If we could have something which runs on the first of the month and updates that lock file it would be feasible to lock all the test dependencies, I think.

**Comment by tillahoffmann** [comment]
Yes, that should be possible with https://github.com/marketplace/actions/create-pull-request.

**Comment by WardBrian** [comment]
This has been less of an issue in the years since, so I think we can officially say we don't need to do this
